Arrowgene.Ddon.GameServer/Characters/JobManager.cs

```diff
@@ -317,6 +405,115 @@ public void RemoveSkill(IDatabase database, CharacterCommon character, JobId job
+                    new uint[] { LearnedNormalSkills.HighScepter.ControlledFall },
+            },
+            [JobId.Hunter] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.Hunter.QuickBowStyle },
+                    new uint[] { LearnedNormalSkills.Hunter.QuickChargeStyle },
+                    new uint[] { LearnedNormalSkills.Hunter.KeenSight }
+            },
+            [JobId.Priest] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.Priest.HealAura },
+                    new uint[] { LearnedNormalSkills.Priest.HolyAura },
+                    new uint[] { LearnedNormalSkills.Priest.Floating }
+            },
+            [JobId.ShieldSage] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.ShieldSage.ElementChange },
+                    new uint[] { LearnedNormalSkills.ShieldSage.ShieldSequence },
+                    new uint[] { LearnedNormalSkills.ShieldSage.Attract }
+            },
+            [JobId.Seeker] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.Seeker.RoundhouseKick },
+                    new uint[] { LearnedNormalSkills.Seeker.HundredKisses, LearnedNormalSkills.Seeker.HundredSlashes },
+                    new uint[] { LearnedNormalSkills.Seeker.DoubleJump }
+            },
+            [JobId.Sorcerer] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.Sorcerer.MagickBoost },
+                    new uint[] { LearnedNormalSkills.Sorcerer.BackMove},
+                    new uint[] { LearnedNormalSkills.Sorcerer.Floating }
+            },
+            [JobId.SpiritLancer] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.SpiritLancer.ThrustMore },
+                    new uint[] { LearnedNormalSkills.SpiritLancer.Aid },
+                    new uint[] { LearnedNormalSkills.SpiritLancer.FallMove }
+            },
+            [JobId.Warrior] = new List<uint[]>() {
+                    new uint[] { LearnedNormalSkills.Warrior.Devastate },
+                    new uint[] { LearnedNormalSkills.Warrior.SavageLash },
+                    new uint[] { LearnedNormalSkills.Warrior.ControlledFall }
+            },
+        };
+
+        public void UnlockLearnedNormalSkill(IDatabase database, GameClient client, CharacterCommon character, JobId Job, uint SkillIndex)
+        {
+            CDataCharacterJobData characterJobData = character.CharacterJobDataList.Where(cjd => cjd.Job == Job).Single();
+
+            if (!gLearnedNormalSkillMap.ContainsKey(Job) || SkillIndex == 0 || ((SkillIndex - 1) > gLearnedNormalSkillMap[Job].Count()))
+            {
+                // Something strange happened, either there is a new job (unlikely)
+                // or there is a missing skill, or someone tried to craft a custom
+                // packet to the server. For now, just return back the same
+                // data requested so the client doesn't hang.
```

Comment:
Rather than failing silently it'd be better to leave a log message and return a response with Error or Result set to some value other than 0 so the player gets kicked to the main menu

Reply:
Ok, this sounds good. Do you have an example? I was looking to do something like this but was not sure how to.